goldmane/pkg/types/flow.go

```diff
@@ -0,0 +1,336 @@
+// Copyright (c) 2025 Tigera, Inc. All rights reserved.
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package types
+
+import (
+	"unique"
+
+	"github.com/sirupsen/logrus"
+
+	"github.com/projectcalico/calico/goldmane/proto"
+	goproto "google.golang.org/protobuf/proto"
+)
+
+type FlowKeySource struct {
+	SourceName      string
+	SourceNamespace string
+	SourceType      proto.EndpointType
+}
+
+type FlowKeyDestination struct {
+	DestName             string
+	DestNamespace        string
```

Comment:
Are we not worried about the overhead of forgoing interning the values here? 

What you've done is definitely the a big chunk, but would there be a downside to interning these values too? The namespaces will at least be a source of constant duplication.

Reply:
I think it's hard to say which approach is going to be more effective at de-duplication - interning the entire structures, or interning the individual values. 

My hunch is that that interning the structures will be - on average - more effective, as it can reduce multiple fields down to a single pointer instead of a pointer per-field, and my suspicion is that the nature of Flow data is such that we will see lots of the same sources and destinations, rather than a full distribution of every combination of name / namespace, etc.

Reply:
I wasn't suggesting that you remove the struct interning, it's a good idea and if it was a one vs the other I agree that the struct interning might be more valuable. 

I'm just wondering if there's any reason why we can't also intern some or all of the strings here as well, since we know they'll be duplicated.

With generics and the unique package, it seems like interning is quite simple to implement. I'm wondering if there's any reason we shouldn't be a bit more liberal with this interning feature in cases where we know the something might be duplicated. 

Unless we know of some issues with the library if it's overused (slight overhead with the internal map lookups?).

Reply:
I'd rather we identify specific areas that would benefit the most from this, and then address them. If profiling comes back and says that the flow keys are still the problem, then we can look at it at as a solution - but the benefit from interning the specific values within the struct is still unclear to me at the moment - a small experiment I ran showed no improvement in doing this, and depending on the nature of the data, interning the individual values could actually be _worse_ than not interning them (although I think it's unlikely we see this show up often in practice).